### ISSUE TITLE
Support for musl/uclibc-ng

### ISSUE THREAD

**Issue by xokdvium** [jeremy-rifkin/cpptrace#128]
Currently cpptrace fails to build for non-glibc standard C libs due to the usage of `dladdr1`, which isn't implemented by musl-libc or uclibc-ng.

For both of those libs it should be possible to use the implementation for apple, which uses `dladdr`.
I see that there's now a fallback for `CPPTRACE_HAS_DL_FIND_OBJECT` to `dladdr1`.
It should be possible to implement a more robust fallback scheme: `_dl_find_object` -> `dladdr1` -> `dladdr` for linux.
This should be enough to support all implementations.

I've tested that the apple implementation works fine for musl by applying the following (very crude) patch:

```patch
diff --git a/src/binary/object.hpp b/src/binary/object.hpp
index 2c812e5..3e988fa 100644
--- a/src/binary/object.hpp
+++ b/src/binary/object.hpp
@@ -22,7 +22,7 @@
 
 namespace cpptrace {
 namespace detail {
-    #if IS_LINUX
+    #if IS_LINUX && !IS_MUSL
     inline std::string resolve_l_name(const char* l_name) {
         if(l_name != nullptr && l_name[0] != 0) {
             return l_name;
@@ -78,7 +78,7 @@ namespace detail {
         return frame;
     }
     #endif
-    #elif IS_APPLE
+    #elif IS_APPLE || IS_MUSL
     // macos doesn't have dladdr1 but it seems its dli_fname behaves more sensibly?
     // dladdr queries are needed to get pre-ASLR addresses and targets to run addr2line on
     inline object_frame get_frame_object_info(frame_ptr address) {
diff --git a/src/utils/common.hpp b/src/utils/common.hpp
index ad4c191..a2c9dfd 100644
--- a/src/utils/common.hpp
+++ b/src/utils/common.hpp
@@ -18,6 +18,21 @@
  #error "Unexpected platform"
 #endif
 
+#define IS_MUSL 0
+#ifndef _GNU_SOURCE
+  #define _GNU_SOURCE
+  #include <features.h>
+  #ifndef __USE_GNU
+    #define IS_MUSL 1
+  #endif
+  #undef _GNU_SOURCE
+#else
+  #include <features.h>
+  #ifndef __USE_GNU
+    #define IS_MUSL 1
+  #endif
+#endif
+
 #define IS_CLANG 0
 #define IS_GCC 0
 #define IS_MSVC 0
```

**Comment by jeremy-rifkin** [comment]
Thanks! I've pushed a change to dev that falls back to normal `dladdr`. I've verified with a musl build that abnormal argv[0] won't cause a problem with `exec -a demo build/demo` and it generated traces correctly. I'm going to do some more research to try to figure out how to best do this fallback safely.

**Comment by jeremy-rifkin** [comment]
I have discovered `dladdr1` was added to glibc in 2003 and I don't think I have to worry about supporting anyone using a glibc more than two decades old.

**Comment by xokdvium** [comment]
I'll hijack this issue a bit. I'd nice to see this fix and https://github.com/jeremy-rifkin/cpptrace/commit/d7c19a5544fb9de405794b4d07b99d0c6e30f579 in the upcoming release, since I'd like to package this library for https://github.com/nixos/nixpkgs. It would be nice to have all the portability issues resolved in a stable release tag.

**Comment by jeremy-rifkin** [comment]
I'm hoping to do a release soon, basically as soon as I can implement #129!

**Comment by jeremy-rifkin** [comment]
Thanks for your patience, I've released v0.6.0
